File: Part 2 (C++)/minJerkTrajectory.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <iomanip>

// Global variables for current position and velocity
std::vector<double> pos = {0.0, 0.0, 0.0}; // Initial position [x, y, z]
std::vector<double> vel = {0.0, 0.0, 0.0}; // Initial velocity [vx, vy, vz]

// Function to calculate minimum jerk trajectory considering initial velocity
void minJerk(const std::vector<double>& start_pos, const std::vector<double>& end_pos, 
             const std::vector<double>& start_vel, double time_step, double segment_time,
             std::vector<std::vector<double>>& new_pos, std::vector<std::vector<double>>& new_vel,
             std::vector<std::vector<double>>& new_acc, std::vector<std::vector<double>>& new_jerk) {
    double T = segment_time; // Duration of the segment
    std::vector<double> tau((int)(T / time_step) + 1);
    for (size_t i = 0; i < tau.size(); ++i) {
        tau[i] = (i * time_step) / T;
    }

    std::vector<double> pos_diff = {end_pos[0] - start_pos[0], end_pos[1] - start_pos[1], end_pos[2] - start_pos[2]};

    new_pos.resize(tau.size(), std::vector<double>(3));
    new_vel.resize(tau.size(), std::vector<double>(3));
    new_acc.resize(tau.size(), std::vector<double>(3));
    new_jerk.resize(tau.size(), std::vector<double>(3));

    for (size_t i = 0; i < tau.size(); ++i) {
        for (int j = 0; j < 3; ++j) {
            new_pos[i][j] = start_pos[j] + start_vel[j] * tau[i] * T + (pos_diff[j] - start_vel[j] * T) * (10*std::pow(tau[i], 3) - 15*std::pow(tau[i], 4) + 6*std::pow(tau[i], 5));
            new_vel[i][j] = start_vel[j] + (pos_diff[j] - start_vel[j] * T) * (30*std::pow(tau[i], 2) - 60*std::pow(tau[i], 3) + 30*std::pow(tau[i], 4)) / T;
            new_acc[i][j] = (pos_diff[j] - start_vel[j] * T) * (60*tau[i] - 180*std::pow(tau[i], 2) + 120*std::pow(tau[i], 3)) / (T * T);
            new_jerk[i][j] = (pos_diff[j] - start_vel[j] * T) * (60 - 360*tau[i] + 360*std::pow(tau[i], 2)) / (T * T * T);
        }
    }
}

int main() {
    // Rectangle vertices in the XY plane
    std::vector<std::vector<double>> rect_vertices = {
        pos,
        {0.1, 0.1, 0},
        {0.5, 0.1, 0},
        {0.5, 0.5, 0},
        {0.1, 0.5, 0},
        {0.1, 0.1, 0}
    };

    // Time settings
    double time_step = 0.008; // 8 milliseconds
    double total_time = 10.0; // Total time for the simulation in seconds
    int num_steps = static_cast<int>(total_time / time_step); // Number of steps

    // Preallocate arrays for storing position, velocity, acceleration, and jerk
    std::vector<std::vector<double>> positions(num_steps, std::vector<double>(3));
    std::vector<std::vector<double>> velocities(num_steps, std::vector<double>(3));
    std::vector<std::vector<double>> accelerations(num_steps, std::vector<double>(3));
    std::vector<std::vector<double>> jerks(num_steps, std::vector<double>(3));

    int step_index = 0; // Index to keep track of the step

    for (size_t i = 0; i < rect_vertices.size() - 1; ++i) {
        // Calculate the current vertex and the next vertex
        std::vector<double> current_vertex = rect_vertices[i];
        std::vector<double> next_vertex = rect_vertices[i + 1];

        // Calculate the minimum jerk trajectory for the current segment
        std::vector<std::vector<double>> segment_pos, segment_vel, segment_acc, segment_jerk;
        minJerk(current_vertex, next_vertex, vel, time_step, total_time / (rect_vertices.size() - 1),
                segment_pos, segment_vel, segment_acc, segment_jerk);

        // Store the results for the current segment
        for (size_t j = 0; j < segment_pos.size(); ++j) {
            if (step_index < num_steps) {
                positions[step_index] = segment_pos[j];
                velocities[step_index] = segment_vel[j];
                accelerations[step_index] = segment_acc[j];
                jerks[step_index] = segment_jerk[j];
                ++step_index;
            }
        }

        // Update the global position and velocity to the end of the current segment
        pos = next_vertex;
        vel = segment_vel.back();
    }

    // Write results to file
    std::ofstream outfile("minJerkTrajectory.txt");
    outfile << std::fixed << std::setprecision(6);
    outfile << "Time;X;Y;Z;Vx;Vy;Vz;Ax;Ay;Az;Jx;Jy;Jz\n";
    for (int i = 0; i < step_index; ++i) {
        double time = i * time_step;
        outfile << time << ";";
        for (int j = 0; j < 3; ++j) outfile << positions[i][j] << ";";
        for (int j = 0; j < 3; ++j) outfile << velocities[i][j] << ";";
        for (int j = 0; j < 3; ++j) outfile << accelerations[i][j] << ";";
        for (int j = 0; j < 3; ++j) outfile << jerks[i][j] << (j < 2 ? ";" : "\n");
    }

    outfile.close();
    return 0;
}

File: Part 1 (MATLAB)/minJerkTrajectory.m
clc; clear; close all;

% Global variables for current position and velocity
global pos vel; %#ok<GVMIS>

% Initial conditions (can be modified)
pos = [0, 0, 0]; % Initial position [x, y, z]
vel = [0, 0, 0]; % Initial velocity [vx, vy, vz]

% Rectangle vertices in the XY plane
rect_vertices = [pos;              % Starting position
                 0.1, 0.1, 0;      % Vertex 1
                 0.5, 0.1, 0;      % Vertex 2
                 0.5, 0.5, 0;      % Vertex 3
                 0.1, 0.5, 0;      % Vertex 4
                 0.1, 0.1, 0];     % Closing the rectangle

% Time settings
time_step = 0.008; % 8 milliseconds
total_time = 10; % Total time for the simulation in seconds
num_steps = total_time / time_step; % Number of steps

% Preallocate arrays for storing position, velocity, acceleration, and jerk
positions = zeros(num_steps, 3);
velocities = zeros(num_steps, 3);
accelerations = zeros(num_steps, 3);
jerks = zeros(num_steps, 3);

% Main loop
step_index = 1; % Index to keep track of the step
for i = 1:length(rect_vertices)-1
    % Calculate the current vertex and the next vertex
    current_vertex = rect_vertices(i, :);
    next_vertex = rect_vertices(i+1, :);
    
    % Calculate the minimum jerk trajectory for the current segment
    [segment_pos, segment_vel, segment_acc, segment_jerk] = minJerk(current_vertex, next_vertex, vel, time_step, total_time / (length(rect_vertices)-1));
    
    % Store the results for the current segment
    for j = 1:length(segment_pos)
        if step_index <= num_steps
            positions(step_index, :) = segment_pos(j, :); % Store position
            velocities(step_index, :) = segment_vel(j, :); % Store velocity
            accelerations(step_index, :) = segment_acc(j, :); % Store acceleration
            jerks(step_index, :) = segment_jerk(j, :); % Store jerk
            step_index = step_index + 1; % Increment step index
        end
    end
    
    % Update the global position and velocity to the end of the current segment
    pos = next_vertex;
    vel = segment_vel(end, :);
end

% Plotting the XY trajectory
figure;
plot(positions(:, 1), positions(:, 2)); % Plot X vs Y positions
title('XY Trajectory');
xlabel('X [m]');
ylabel('Y [m]');
axis equal; % Ensuring the aspect ratio is equal for x and y

% Plotting results in a separate figure
figure;
subplot(4, 1, 1);
plot((0:num_steps-1)*time_step, positions); % Plot positions over time
title('Position');
xlabel('Time [s]');
ylabel('Position [m]');
legend('x', 'y', 'z');

subplot(4, 1, 2);
plot((0:num_steps-1)*time_step, velocities); % Plot velocities over time
title('Velocity');
xlabel('Time [s]');
ylabel('Velocity [m/s]');
legend('vx', 'vy', 'vz');

subplot(4, 1, 3);
plot((0:num_steps-1)*time_step, accelerations); % Plot accelerations over time
title('Acceleration');
xlabel('Time [s]');
ylabel('Acceleration [m/s^2]');
legend('ax', 'ay', 'az');

subplot(4, 1, 4);
plot((0:num_steps-1)*time_step, jerks); % Plot jerks over time
title('Jerk');
xlabel('Time [s]');
ylabel('Jerk [m/s^3]');
legend('jx', 'jy', 'jz');

% Function to calculate minimum jerk trajectory considering initial velocity
function [new_pos, new_vel, new_acc, new_jerk] = minJerk(start_pos, end_pos, start_vel, time_step, segment_time)
    T = segment_time; % Duration of the segment
    tau = (0:time_step:T)'/T; % Normalized time vector
    
    pos_diff = end_pos - start_pos; % Position difference
    vel_diff = -start_vel; %#ok<NASGU> % Velocity difference (initial velocity assumption)
    
    % Minimum jerk position trajectory considering initial velocity
    new_pos = start_pos + start_vel .* tau * T + (pos_diff - start_vel * T) .* (10*tau.^3 - 15*tau.^4 + 6*tau.^5);
    
    % Minimum jerk velocity trajectory considering initial velocity
    new_vel = start_vel + (pos_diff - start_vel * T) .* (30*tau.^2 - 60*tau.^3 + 30*tau.^4) / T;
    
    % Minimum jerk acceleration trajectory
    new_acc = (pos_diff - start_vel * T) .* (60*tau - 180*tau.^2 + 120*tau.^3) / T^2;
    
    % Minimum jerk jerk trajectory
    new_jerk = (pos_diff - start_vel * T) .* (60 - 360*tau + 360*tau.^2) / T^3;
end
